### PR TITLE
add more test coverage tonative  hmac sign/verify tests

### PR DESCRIPTION
**Pull request by meilu-git**

### CODE REVIEW

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 1]
```diff
@@ -112,6 +113,162 @@ struct HmacTestParams
...
+
+    std::vector<uint8_t> signature(sig_buf.len);
+    sig_buf.ptr = signature.data();
+
+    auto sign_err = azihsm_crypt_sign(&algo, hmac_key, &data_buf, &sig_buf);
+    EXPECT_EQ(sign_err, AZIHSM_STATUS_SUCCESS);
+    EXPECT_GT(sig_buf.len, 0);
+
+    signature.resize(sig_buf.len);
+    return signature;
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp:202]: resolved

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 2]
```diff
@@ -112,6 +113,162 @@ struct HmacTestParams
...
+    azihsm_handle hmac_key,
+    azihsm_algo &algo,
+    const std::vector<uint8_t> &msg,
+    const std::vector<size_t> &chunk_sizes
+)
+{
+    auto_ctx sign_ctx;
+
+    EXPECT_EQ(azihsm_crypt_sign_init(&algo, hmac_key, sign_ctx.get_ptr()), AZIHSM_STATUS_SUCCESS);
+
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp:226]: resolved

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 3]
```diff
@@ -112,6 +113,162 @@ struct HmacTestParams
...
+    azihsm_buffer sig_buf = { .ptr = nullptr, .len = 0 };
+
+    EXPECT_EQ(azihsm_crypt_sign_finish(sign_ctx, &sig_buf), AZIHSM_STATUS_BUFFER_TOO_SMALL);
+    EXPECT_GT(sig_buf.len, 0);
+
+    std::vector<uint8_t> signature(sig_buf.len);
+    sig_buf.ptr = signature.data();
+
+    EXPECT_EQ(azihsm_crypt_sign_finish(sign_ctx, &sig_buf), AZIHSM_STATUS_SUCCESS);
+    EXPECT_GT(sig_buf.len, 0);
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp]: resolved

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 4]
```diff
@@ -112,6 +113,162 @@ struct HmacTestParams
...
+    {
+        const size_t requested_chunk_size = chunk_sizes[chunk_index++ % chunk_sizes.size()];
+        const size_t end = std::min(offset + requested_chunk_size, msg.size());
+
+        azihsm_buffer chunk_buf = {
+            .ptr = const_cast<uint8_t *>(msg.data() + offset),
+            .len = static_cast<uint32_t>(end - offset),
+        };
+
+        EXPECT_EQ(azihsm_crypt_verify_update(verify_ctx, &chunk_buf), AZIHSM_STATUS_SUCCESS);
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp]: resolved

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 5]
```diff
@@ -206,4 +363,1300 @@ TEST_F(azihsm_hmac_sign_verify, sign_verify_hmac_streaming_all_algorithms)
...
+          AZIHSM_ALGO_ID_HMAC_SHA384,
+          AZIHSM_ECC_CURVE_P384,
+          "SHA384" },
+        { AZIHSM_KEY_KIND_HMAC_SHA512,
+          AZIHSM_ALGO_ID_HMAC_SHA512,
+          AZIHSM_ECC_CURVE_P521,
+          "SHA512" },
+    };
+
+    for (const auto &test_case : test_cases)
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp]: resolved

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 6]
```diff
@@ -206,4 +363,1300 @@ TEST_F(azihsm_hmac_sign_verify, sign_verify_hmac_streaming_all_algorithms)
...
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp]: resolved

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 7]
```diff
@@ -206,4 +363,1300 @@ TEST_F(azihsm_hmac_sign_verify, sign_verify_hmac_streaming_all_algorithms)
...
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp]: resolved

[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp]: resolved

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 8]
```diff
@@ -206,4 +363,1300 @@ TEST_F(azihsm_hmac_sign_verify, sign_verify_hmac_streaming_all_algorithms)
...
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp]: resolved

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 9]
```diff
@@ -206,4 +363,1300 @@ TEST_F(azihsm_hmac_sign_verify, sign_verify_hmac_streaming_all_algorithms)
...
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp]: resolved

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 10]
```diff
@@ -206,4 +363,1300 @@ TEST_F(azihsm_hmac_sign_verify, sign_verify_hmac_streaming_all_algorithms)
...
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp]: resolved

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 11]
```diff
@@ -112,6 +113,162 @@ struct HmacTestParams
...
+    const std::vector<size_t> &chunk_sizes,
+    const std::vector<uint8_t> &tag
+)
+{
+    auto_ctx verify_ctx;
+
+    EXPECT_EQ(
+        azihsm_crypt_verify_init(&algo, hmac_key, verify_ctx.get_ptr()),
+        AZIHSM_STATUS_SUCCESS
+    );
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp]: resolved

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 12]
```diff
@@ -197,13 +370,1247 @@ TEST_F(azihsm_hmac_sign_verify, sign_verify_hmac_streaming_all_algorithms)
...
+
+            std::vector<uint8_t> empty_data;
+
+            azihsm_algo algo = {
+                .id = test_case.algo_id,
+                .params = nullptr,
+                .len = 0,
+            };
+
+            test_single_shot_hmac_sign_verify(hmac_key.get(), algo, empty_data);
```
Comment (rajesh-gali) [inline api/tests/cpp/algo/hmac/sign_verify_tests.cpp:633]: can we also add a test to validate null ptr message

Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp:633]: added

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 13]
```diff
@@ -129,13 +129,202 @@ TEST_F(azihsm_hmac_sign_verify, sign_verify_hmac_all_algorithms)
...
+
+    std::vector<uint8_t> signature(sig_buf.len);
+    sig_buf.ptr = signature.data();
+
+    auto sign_err = azihsm_crypt_sign(&algo, hmac_key, &data_buf, &sig_buf);
+    EXPECT_EQ(sign_err, AZIHSM_STATUS_SUCCESS);
+    EXPECT_GT(sig_buf.len, 0);
+
+    signature.resize(sig_buf.len);
+    return signature;
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp:202]: resloved

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 14]
```diff
@@ -197,13 +388,1310 @@ TEST_F(azihsm_hmac_sign_verify, sign_verify_hmac_streaming_all_algorithms)
...
+
+            azihsm_buffer sig_buf = { .ptr = nullptr, .len = 0 };
+            ASSERT_EQ(
+                azihsm_crypt_sign(&algo, signing_hmac_key.get(), &data_buf, &sig_buf),
+                AZIHSM_STATUS_BUFFER_TOO_SMALL
+            );
+
+            std::vector<uint8_t> signature(sig_buf.len);
+            sig_buf.ptr = signature.data();
+
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp]: resolved per suggestions.

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 15]
```diff
@@ -197,13 +388,1310 @@ TEST_F(azihsm_hmac_sign_verify, sign_verify_hmac_streaming_all_algorithms)
...
+        };
+
+        ASSERT_EQ(azihsm_crypt_sign_update(sign_ctx, &msg_buf), AZIHSM_STATUS_SUCCESS);
+
+        azihsm_buffer sig_buf = { .ptr = nullptr, .len = 0 };
+        ASSERT_EQ(azihsm_crypt_sign_finish(sign_ctx, &sig_buf), AZIHSM_STATUS_BUFFER_TOO_SMALL);
+
+        std::vector<uint8_t> tag(sig_buf.len);
+        sig_buf.ptr = tag.data();
+
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp]: resolved per suggestions.

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 16]
```diff
@@ -197,13 +388,1310 @@ TEST_F(azihsm_hmac_sign_verify, sign_verify_hmac_streaming_all_algorithms)
...
+            };
+
+            azihsm_buffer sig_buf = { .ptr = nullptr, .len = 0 };
+            ASSERT_EQ(
+                azihsm_crypt_sign(&algo, hmac_key.get(), &data_buf, &sig_buf),
+                AZIHSM_STATUS_BUFFER_TOO_SMALL
+            );
+
+            std::vector<uint8_t> signature(sig_buf.len);
+            sig_buf.ptr = signature.data();
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp:719]: resolved per suggestions.

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 17]
```diff
@@ -197,13 +398,1310 @@ TEST_F(azihsm_hmac_sign_verify, sign_verify_hmac_streaming_all_algorithms)
...
+                AZIHSM_STATUS_SUCCESS
+            );
+
+            ASSERT_GT(sig_buf.len, 0);
+            signature[0] ^= 0x01;
+
+            ASSERT_NE(
+                azihsm_crypt_verify(&algo, hmac_key.get(), &data_buf, &sig_buf),
+                AZIHSM_STATUS_SUCCESS
+            );
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp]: resolved.

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 18]
```diff
@@ -112,10 +113,9 @@ struct HmacTestParams
...
 };
 
-// HMAC Single-Shot Sign/Verify Tests
-TEST_F(azihsm_hmac_sign_verify, sign_verify_hmac_all_algorithms)
+static std::vector<HmacTestParams> hmac_test_cases()
 {
-    std::vector<HmacTestParams> test_cases = {
+    return {
         { AZIHSM_KEY_KIND_HMAC_SHA256,
           AZIHSM_ALGO_ID_HMAC_SHA256,
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp:120]: it's already rixed.

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 19]
```diff
@@ -197,13 +398,1310 @@ TEST_F(azihsm_hmac_sign_verify, sign_verify_hmac_streaming_all_algorithms)
...
+
+            azihsm_buffer tampered_data_buf = {
+                .ptr = data.data(),
+                .len = static_cast<uint32_t>(data.size()),
+            };
+
+            ASSERT_NE(
+                azihsm_crypt_verify(&algo, hmac_key.get(), &tampered_data_buf, &sig_buf),
+                AZIHSM_STATUS_SUCCESS
+            );
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp]: resolved per suggestions.

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 20]
```diff
@@ -197,13 +398,1310 @@ TEST_F(azihsm_hmac_sign_verify, sign_verify_hmac_streaming_all_algorithms)
...
+            sig_buf.ptr = signature.data();
+            ASSERT_EQ(
+                azihsm_crypt_sign(&algo, signing_hmac_key.get(), &data_buf, &sig_buf),
+                AZIHSM_STATUS_SUCCESS
+            );
+
+            ASSERT_NE(
+                azihsm_crypt_verify(&algo, verifying_hmac_key.get(), &data_buf, &sig_buf),
+                AZIHSM_STATUS_SUCCESS
+            );
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp]: resolved per suggestions.

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 21]
```diff
@@ -197,13 +398,1310 @@ TEST_F(azihsm_hmac_sign_verify, sign_verify_hmac_streaming_all_algorithms)
...
+
+            ASSERT_GT(sig_buf.len, 1);
+            sig_buf.len -= 1;
+
+            ASSERT_NE(
+                azihsm_crypt_verify(&algo, hmac_key.get(), &data_buf, &sig_buf),
+                AZIHSM_STATUS_SUCCESS
+            );
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp]: resolved per suggestions.

File: api/tests/cpp/algo/hmac/sign_verify_tests.cpp [hunk 22]
```diff
@@ -197,13 +398,1310 @@ TEST_F(azihsm_hmac_sign_verify, sign_verify_hmac_streaming_all_algorithms)
...
+
+            ASSERT_NE(
+                azihsm_crypt_verify(&algo, hmac_key.get(), &data_buf, &truncated_buf),
+                AZIHSM_STATUS_SUCCESS
+            );
+
+            ASSERT_NE(
+                azihsm_crypt_verify(&algo, hmac_key.get(), &data_buf, &extended_buf),
+                AZIHSM_STATUS_SUCCESS
+            );
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/cpp/algo/hmac/sign_verify_tests.cpp]: resolved per suggestions.

**Review by rajesh-gali — APPROVED** [review]
